### PR TITLE
Add Nesterov Momentum

### PR DESCRIPTION
**Pull request by apaszke**
Adds the Nesterov variant of momentum as an option for SGD.

By @ajbrock. I rebased the changes and added `__setstate__`.

### CODE REVIEW

File: torch/optim/sgd.py [hunk 1]
```diff
@@ -3,31 +3,37 @@
...
         >>> optimizer.step()
     """
 
     def __init__(self, params, lr=required, momentum=0, dampening=0,
-                 weight_decay=0):
+                 weight_decay=0, nesterov=False):
         defaults = dict(lr=lr, momentum=momentum, dampening=dampening,
-                        weight_decay=weight_decay)
+                        weight_decay=weight_decay, nesterov=nesterov)
+        if nesterov and (momentum <= 0 and dampening != 0):
```
Comment (colesbury) [inline torch/optim/sgd.py]: should this be `momentum <= 0 **or** dampening != 0`?

File: torch/optim/sgd.py [hunk 2]
```diff
@@ -3,31 +3,37 @@
...
     """Implements stochastic gradient descent (optionally with momentum).
-
     Args:
         params (iterable): iterable of parameters to optimize or dicts defining
             parameter groups
         lr (float): learning rate
         momentum (float, optional): momentum factor (default: 0)
         weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+        dampening (float, optional): dampening for momentum (default: 0)
+        nesterov(bool, optional): enables Nesterov momentum (default: False)
```
Comment (colesbury) [inline torch/optim/sgd.py]: space between `nesterov` and `(`

**Review by colesbury — COMMENTED** [review]
Since the formulation is different from how Nesterov accelerated gradient is traditionally written, maybe we should mention that in a code or doc comment.

I think @soumith has a link to the (Ilya Sutskever's?) document that derives this formulation.

File: torch/optim/sgd.py (hunk 1 above)
Comment (ajbrock) [inline torch/optim/sgd.py]: I pulled this one straight from [the legacy version](https://github.com/pytorch/pytorch/blob/master/torch/legacy/optim/sgd.py#L43), but I think you're correct that it should be 0 so that the error is raised if either momentum is non-positive **or** dampening is non-zero, yeah.

Comment (apaszke) [inline torch/optim/sgd.py]: Yeah it should

**Comment by ajbrock** [comment]
The formulation is described [here](https://github.com/lisa-lab/pylearn2/pull/136#issuecomment-10381617), and I think it's equivalent to Sutskever's formulation [in equation 3 of the "importance of init and momentum" paper](http://www.cs.toronto.edu/~hinton/absps/momentum.pdf).

**Comment by apaszke** [comment]
I've added a link to Ivo Danihelka's pdf from torch/optim#27.

**Comment by vadimkantorov** [comment]
The link to PDF from https://github.com/torch/optim/issues/27#issuecomment-58915310: https://github.com/fidlej/optim/raw/master/dok/nesterov_simple.pdf seems currently missing from Nesterov docs in PyTorch. Moreover, the docs a bit confuse the reader by mentioning that update follows Sutskever paper, but the paper uses more traditional look-ahead formulation.

More context:
https://github.com/pytorch/pytorch/pull/5920#issuecomment-375181908
https://github.com/lisa-lab/pylearn2/pull/136#issuecomment-10381617
https://github.com/pytorch/pytorch/pull/887#issuecomment-283400174

It would be cool to have these derivations in the doc or at least in comments (also small pdf notes can sometimes disappear from internet, because someone's university homepage is moved or deleted)

cc @apaszke
